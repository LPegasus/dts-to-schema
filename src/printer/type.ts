import type * as ts from "typescript";

export interface PrintAdaptor {
  emitSchema(printSourceData: PrintSourceData): Promise<void>;
  emitEntry(
    outputFileName: string,
    printSourceDataCollection: PrintSourceData[]
  ): Promise<void>;
}

export enum DtsModuleType {
  Legacy = "legacy",
  Module = "module",
}

export type PrintSourceData = PrintSourceEnumData | PrintSourceInterfaceData;

export interface PrintSourceDataBase {
  /** 输出目录，输出目录通过 namespace + outDir 决定 */
  outputDir: string;
  moduleType: DtsModuleType;
  fullyQualifiedName: string;
  /** TS 定义所在文件名，生成的 schema 文件需要类型定义 */
  relativeDefinitionFileName?: string;
  /** interface 所有成员，string 表示是原始类型，PrintTypeReference 表示用 ref */
  kind: ts.SyntaxKind;
}

export interface PrintSourceEnumData extends PrintSourceDataBase {
  kind: ts.SyntaxKind.EnumDeclaration;
  members: Array<{ name: string; value: string | number }>;
}

export interface PrintSourceInterfaceData extends PrintSourceDataBase {
  kind: ts.SyntaxKind.InterfaceDeclaration;
  props: Record<string, PrintMemberType>;
}

export interface BasePrintMemberType {
  v: string | PrintTypeReference | "any";
  optional: boolean;
  nullable: boolean;
}

export interface PrintMemberTypeOfArray extends BasePrintMemberType {
  isArray: true;
  latitude: number;
}

export interface PrintMemberTypeOfObject extends BasePrintMemberType {
  isArray: false;
}

export type PrintMemberType = PrintMemberTypeOfArray | PrintMemberTypeOfObject;

export interface PrintTypeReference {
  ref: string;
}

export enum PrinterEmitSchemaType {
  JTD = "jtd",
  JSON = "json",
}

export interface PrinterConfiguration {
  /** output directory */
  outDir?: string;
  emitSchemaType?: PrinterEmitSchemaType;
  prettierConfig?: import("prettier").Options;
}

export interface PrintHost {
  writeFileAsync(fileName: string, content: string): Promise<void>;
  getOutputAbsoluteFileName(printSourceData: PrintSourceData): string;
  normalizeVariableNameFromQualifiedName(fullyQualifiedName: string): string;
}
