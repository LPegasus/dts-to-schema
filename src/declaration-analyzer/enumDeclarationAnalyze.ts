import * as ts from "typescript";
import {
  DtsModuleType,
  PrintSourceData,
  PrintSourceEnumData,
} from "../printer/type";
import { baseAnalyze } from "./baseAnalyze";
import * as path from "path";

export function enumDeclarationAnalyze(
  typeChecker: ts.TypeChecker,
  node: ts.EnumDeclaration,
  outputDir: string,
  accumulator: Map<string, PrintSourceData>
): PrintSourceEnumData {
  const { fullyQualifiedName, lastDeclaration } = baseAnalyze(
    node,
    typeChecker
  );
  if (accumulator.has(fullyQualifiedName)) {
    return accumulator.get(fullyQualifiedName)! as any;
  }

  const splittedQualifiedNames = fullyQualifiedName.split(".");
  const emitDir = path.resolve(
    outputDir,
    ...splittedQualifiedNames.slice(0, -1)
  );

  let incrementalToken = 0;

  const printSourceData: PrintSourceData = {
    kind: node.kind,
    fullyQualifiedName,
    outputDir: emitDir,
    relativeDefinitionFileName: path.relative(
      outputDir,
      lastDeclaration.getSourceFile().fileName
    ),
    moduleType: DtsModuleType.Legacy,
    members: node.members.map((member) => {
      const { initializer, name } = member;
      let value: any;
      if (initializer === undefined) {
        value = incrementalToken++;
      } else if (ts.isStringLiteral(initializer)) {
        value = initializer.getText();
      } else if (ts.isNumericLiteral(initializer)) {
        value = Number(initializer.getText());
        incrementalToken = value + 1;
      }
      return {
        name: (name as ts.Identifier).getText() as string,
        value,
      };
    }),
  };

  return printSourceData;
}
