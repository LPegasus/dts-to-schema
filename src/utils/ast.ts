import * as ts from "typescript";

export function printImportStartStatement(
  aliasName: string,
  moduleSpecifier: string
): string {
  const ast = ts.factory.createImportDeclaration(
    undefined,
    undefined,
    ts.factory.createImportClause(
      false,
      undefined,
      ts.factory.createNamespaceImport(ts.factory.createIdentifier(aliasName))
    ),
    ts.factory.createStringLiteral(moduleSpecifier),
    undefined
  );

  const code = ts
    .createPrinter({ newLine: ts.NewLineKind.LineFeed })
    .printNode(
      ts.EmitHint.Expression,
      ast,
      ts.createSourceFile("", "", ts.ScriptTarget.ESNext, false)
    );

  return code;
}

export function filterNodeForSourceFile(
  sourceFile: ts.SourceFile,
  predicate: (node: ts.Node) => boolean
) {
  const results: ts.Node[] = [];
  const scan = (node: ts.Node) => {
    if (predicate(node)) {
      results.push(node);
    }
    node.forEachChild(scan);
  };

  scan(sourceFile);

  return results;
}
