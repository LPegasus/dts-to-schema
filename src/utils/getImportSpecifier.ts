import type { DtsModuleType } from "../printer/type";

export function getImportSpecifier(
  fileName: string,
  dtsModuleType: DtsModuleType
) {}
