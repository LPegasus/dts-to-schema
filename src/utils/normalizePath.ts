export function normalizePath(path: string) {
  let p = path.replace(/\\/g, "/");

  if (process.platform === "win32") {
    p = p.replace(/^([a-zA-Z]:)/, "/$1");
  }

  if (!/^[./]/.test(p)) {
    p = "./" + p;
  }
  return p;
}
