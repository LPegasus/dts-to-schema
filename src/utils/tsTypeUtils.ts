import * as ts from "typescript";

export function getEnumValueType(node: ts.EnumDeclaration): ts.SyntaxKind[] {
  return node.members.map((member) => {
    return member.initializer?.kind ?? ts.SyntaxKind.NumericLiteral;
  });
}

export function isEnumMembersAllNumber(node: ts.EnumDeclaration): boolean {
  return getEnumValueType(node).every(
    (type) => type === ts.SyntaxKind.NumericLiteral
  );
}
