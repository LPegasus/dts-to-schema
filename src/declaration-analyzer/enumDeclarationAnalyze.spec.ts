import { enumDeclarationAnalyze } from "./enumDeclarationAnalyze";
import * as ts from "typescript";
import { filterNodeForSourceFile } from "../utils/ast";
import type { PrintSourceData } from "../printer/type";

const compilerOptions = {
  module: ts.ModuleKind.ESNext,
  moduleResolution: ts.ModuleResolutionKind.NodeJs,
  target: ts.ScriptTarget.ESNext,
  types: [],
};
const compilerHost = ts.createCompilerHost(compilerOptions, true);

compilerHost.fileExists = () => true;
compilerHost.readFile = (fileName: string) => {
  const testSnaps: Record<string, string> = {
    "Foo.d.ts": `declare namespace Foo {
    enum FooEnum {
      A = "A",
      B = 0,
      C
    };
  }`,
    "Bar.d.ts": `enum BarEnum {
    A = 0x01,
    B = 0x03,
    C = 0x05
  };`,
  };
  return testSnaps[fileName];
};

describe("enum declaration analyze", () => {
  let prog: ts.Program;
  let typeChecker: ts.TypeChecker;

  beforeEach(() => {
    prog = ts.createProgram({
      rootNames: ["Foo.d.ts", "Bar.d.ts"],
      options: {
        module: ts.ModuleKind.ESNext,
        moduleResolution: ts.ModuleResolutionKind.NodeJs,
        target: ts.ScriptTarget.ESNext,
        types: [],
      },
      host: compilerHost,
    });

    typeChecker = prog.getTypeChecker();
  });

  test("Foo.d.ts", () => {
    const sourceFile = prog.getSourceFile("Foo.d.ts")!;
    const enumNodeList = filterNodeForSourceFile(sourceFile, (n) =>
      ts.isEnumDeclaration(n)
    );
    expect(enumNodeList.length).toEqual(1);

    const cache = new Map<string, PrintSourceData>();
    const printSourceData = enumDeclarationAnalyze(
      typeChecker,
      enumNodeList[0] as ts.EnumDeclaration,
      "",
      cache
    );

    expect(printSourceData.fullyQualifiedName).toEqual("Foo.FooEnum");
    expect(printSourceData.members).toMatchSnapshot();
  });

  test("Bar.d.ts", () => {
    const sourceFile = prog.getSourceFile("Bar.d.ts")!;
    const enumNodeList = filterNodeForSourceFile(sourceFile, (n) =>
      ts.isEnumDeclaration(n)
    );
    expect(enumNodeList.length).toEqual(1);

    const cache = new Map<string, PrintSourceData>();
    const printSourceData = enumDeclarationAnalyze(
      typeChecker,
      enumNodeList[0] as ts.EnumDeclaration,
      "",
      cache
    );

    expect(printSourceData.members).toMatchSnapshot();
  });
});
