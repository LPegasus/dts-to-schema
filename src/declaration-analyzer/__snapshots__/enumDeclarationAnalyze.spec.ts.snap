// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enum declaration analyze Bar.d.ts 1`] = `
Array [
  Object {
    "name": "A",
    "value": 1,
  },
  Object {
    "name": "B",
    "value": 3,
  },
  Object {
    "name": "C",
    "value": 5,
  },
]
`;

exports[`enum declaration analyze Foo.d.ts 1`] = `
Array [
  Object {
    "name": "A",
    "value": "\\"A\\"",
  },
  Object {
    "name": "B",
    "value": 0,
  },
  Object {
    "name": "C",
    "value": 1,
  },
]
`;
